p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,350 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/network"
+
+	tpt "github.com/libp2p/go-libp2p-core/transport"
+	"github.com/libp2p/go-libp2p/p2p/security/noise"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/multiformats/go-multibase"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/pion/dtls/v2/pkg/crypto/fingerprint"
+	"github.com/pion/ice/v2"
+	"github.com/pion/webrtc/v3"
+)
+
+var (
+	ErrDataChannelTimeout    = errors.New("timed out waiting for datachannel")
+	ErrNoiseHandshakeTimeout = errors.New("noise handshake timeout")
+	ErrNoCertInConfig        = errors.New("no certificate configured in listener config")
+)
+
+var (
+	defaultMultihash *multihash.DecodedMultihash = nil
+)
+
+func init() {
+	// populate default multihash
+	encoded, err := hex.DecodeString("ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad")
+	if err != nil {
+		panic(err)
+	}
+
+	defaultMultihash = &multihash.DecodedMultihash{
+		Code:   multihash.SHA2_256,
+		Name:   multihash.Codes[multihash.SHA2_256],
+		Digest: encoded,
+		Length: len(encoded),
+	}
+
+}
+
+/// implement net.Listener
+type listener struct {
+	transport                 *WebRTCTransport
+	config                    webrtc.Configuration
+	privKey                   crypto.PrivKey
+	localFingerprint          webrtc.DTLSFingerprint
+	localFingerprintMultibase string
+	mux                       *UDPMuxNewAddr
+	closeChan                 chan struct{}
+	localMultiaddr            ma.Multiaddr
+	connChan                  chan tpt.CapableConn
+}
+
+func newListener(transport *WebRTCTransport, laddr ma.Multiaddr, socket net.PacketConn, privKey crypto.PrivKey, config webrtc.Configuration) (*listener, error) {
+	mux := NewUDPMuxNewAddr(ice.UDPMuxParams{UDPConn: socket}, make(chan candidateAddr, 1))
+	localFingerprints, err := config.Certificates[0].GetFingerprints()
+	if err != nil {
+		return nil, err
+	}
+
+	localMh, err := hex.DecodeString(strings.ReplaceAll(localFingerprints[0].Value, ":", ""))
+	if err != nil {
+		return nil, err
+	}
+	localMhBuf, _ := multihash.EncodeName(localMh, sdpHashToMh[localFingerprints[0].Algorithm])
+	localFpMultibase, _ := multibase.Encode(multibase.Base64, localMhBuf)
+
+	l := &listener{
+		mux:                       mux,
+		transport:                 transport,
+		privKey:                   privKey,
+		config:                    config,
+		localFingerprint:          localFingerprints[0],
+		localFingerprintMultibase: localFpMultibase,
+		closeChan:                 make(chan struct{}, 1),
+		localMultiaddr:            laddr,
+		connChan:                  make(chan tpt.CapableConn, 20),
+	}
+	go l.startAcceptLoop()
```

Comment:
This will need some more careful shutdown logic, using a `sync.WaitGroup`. When `Close` returns, we need to be sure that no Go routines are running any more. Take a look at https://github.com/libp2p/go-libp2p/blob/7facd81bba8889ee33ac6b4235cb569e652e32eb/p2p/protocol/holepunch/svc.go#L87-L88 for an example.